File: 프로그래머스/파이썬/Lv3/베스트앨범.py
from collections import defaultdict


def solution(genres, plays):
    answer = []

    check = defaultdict(list)
    _sum = defaultdict(int)

    for i in range(len(genres)):
        genre = genres[i]

        if _sum[genre]:
            _sum[genre] += plays[i]
        else:
            _sum[genre] = plays[i]

        check[genre].append([plays[i], i])

    _sum_item = list(_sum.items())
    _sum_item.sort(key=lambda x: -x[1])

    for g, value in _sum_item:
        g_list = check[g]
        if len(g_list) == 1:
            answer.append(g_list[0][1])
            continue
        g_list.sort(key=lambda x: -x[0])

        for i in range(2):
            answer.append(g_list[i][1])

    return answer

File: 프로그래머스/JavaScript/Lv3/베스트앨범.js
function solution(genres, plays) {
    var answer = [];
    const n = genres.length;
    const songIdx = {};
    const totalCount = {};
    
    for (let i = 0; i < n; i++){
        const genre = genres[i]
        const play = plays[i]
        
        if (totalCount[genre]) totalCount[genre] += play
        else totalCount[genre] = play
        
        if (songIdx[genre]) {
            const [firstPlay, firstIdx] = songIdx[genre][0]
            const [secondPlay, secondIdx] = songIdx[genre][1]
            
            if (firstPlay < play || (firstPlay === play && firstIdx > i)) {
                songIdx[genre][0] = [play, i]
                songIdx[genre][1] = [firstPlay, firstIdx]
            } else if (secondPlay < play || (secondPlay === play && secondIdx > i)) {
                songIdx[genre][1] = [play, i]
            }
        } else {
            songIdx[genre] = [[play, i], [0, 0]]
        }
    }
    
    const countItem = Object.entries(totalCount).sort((a, b) => b[1] - a[1])

    for (let i = 0; i < countItem.length; i++){
        const song = countItem[i][0]
        for (let j = 0; j < 2; j++){
            if (songIdx[song][j][0] > 0) answer.push(songIdx[song][j][1])
        }
    }
    return answer;
}